# Challenge 1 - Tic Tac Toe

In this lab you will perform deep learning analysis on a dataset of playing [Tic Tac Toe](https://en.wikipedia.org/wiki/Tic-tac-toe).

There are 9 grids in Tic Tac Toe that are coded as the following picture shows:

![Tic Tac Toe Grids](../images/tttboard.jpg)

In the first 9 columns of the dataset you can find which marks (`x` or `o`) exist in the grids. If there is no mark in a certain grid, it is labeled as `b`. The last column is `class` which tells you whether Player X (who always moves first in Tic Tac Toe) wins in this configuration. Note that when `class` has the value `False`, it means either Player O wins the game or it ends up as a draw.

Follow the steps suggested below to conduct a neural network analysis using Tensorflow and Keras. You will build a deep learning model to predict whether Player X wins the game or not.

## Step 1: Data Engineering

This dataset is almost in the ready-to-use state so you do not need to worry about missing values and so on. Still, some simple data engineering is needed.

1. Read `tic-tac-toe.csv` into a dataframe.
1. Inspect the dataset. Determine if the dataset is reliable by eyeballing the data.
1. Convert the categorical values to numeric in all columns.
1. Separate the inputs and output.
1. Normalize the input data.

In [1]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn import preprocessing
from sklearn.model_selection import train_test_split
from keras import models
from keras.models import Sequential
from keras.layers import Dense, Activation
from sklearn.metrics import r2_score
from sklearn.metrics import confusion_matrix
from sklearn.metrics import accuracy_score

Using TensorFlow backend.


In [2]:
df = pd.read_csv('..//data//tic-tac-toe.csv')

In [3]:
df.head()

TL TM TR ML MM MR BL BM BR  class
0  x  x  x  x  o  o  x  o  o   True
1  x  x  x  x  o  o  o  x  o   True
2  x  x  x  x  o  o  o  o  x   True
3  x  x  x  x  o  o  o  b  b   True
4  x  x  x  x  o  o  b  o  b   True

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 958 entries, 0 to 957
Data columns (total 10 columns):
TL       958 non-null object
TM       958 non-null object
TR       958 non-null object
ML       958 non-null object
MM       958 non-null object
MR       958 non-null object
BL       958 non-null object
BM       958 non-null object
BR       958 non-null object
class    958 non-null bool
dtypes: bool(1), object(9)
memory usage: 68.4+ KB


In [5]:
df['class'].value_counts()

True     626
False    332
Name: class, dtype: int64

In [6]:
columns = list(df.columns)
columns.remove('class')

In [7]:
for i in columns:
    df.replace({i: {'x': 2, 'o': 1, 'b':0}}, inplace = True)

lb = preprocessing.LabelBinarizer()
df['class'] = lb.fit_transform(df['class'])

In [8]:
df.head()

TL  TM  TR  ML  MM  MR  BL  BM  BR  class
0   2   2   2   2   1   1   2   1   1      1
1   2   2   2   2   1   1   1   2   1      1
2   2   2   2   2   1   1   1   1   2      1
3   2   2   2   2   1   1   1   0   0      1
4   2   2   2   2   1   1   0   1   0      1

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 958 entries, 0 to 957
Data columns (total 10 columns):
TL       958 non-null int64
TM       958 non-null int64
TR       958 non-null int64
ML       958 non-null int64
MM       958 non-null int64
MR       958 non-null int64
BL       958 non-null int64
BM       958 non-null int64
BR       958 non-null int64
class    958 non-null int32
dtypes: int32(1), int64(9)
memory usage: 71.2 KB


## Step 2: Build Neural Network

To build the neural network, you can refer to your own codes you wrote while following the [Deep Learning with Python, TensorFlow, and Keras tutorial](https://www.youtube.com/watch?v=wQ8BIBpya2k) in the lesson. It's pretty similar to what you will be doing in this lab.

1. Split the training and test data.
1. Create a `Sequential` model.
1. Add several layers to your model. Make sure you use ReLU as the activation function for the middle layers. Use Softmax for the output layer because each output has a single lable and all the label probabilities add up to 1.
1. Compile the model using `adam` as the optimizer and `sparse_categorical_crossentropy` as the loss function. For metrics, use `accuracy` for now.
1. Fit the training data.
1. Evaluate your neural network model with the test data.
1. Save your model as `tic-tac-toe.model`.

In [10]:
y = pd.DataFrame(df['class'])
X = df.drop(columns=['class']).copy()

In [11]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)

In [12]:
# Build neural network
df_model = models.Sequential()
df_model.add(Dense(units=162, kernel_initializer='uniform', activation='relu', input_dim=9))
df_model.add(Dense(units=162, kernel_initializer='uniform', activation='relu'))
df_model.add(Dense(units=1, kernel_initializer='uniform', activation='sigmoid'))

# Compile model
df_model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

# Train model
df_model.fit(X_train, y_train, batch_size=2, epochs=50)

print('\n######################################################################################\n')
val_loss, val_acc = df_model.evaluate(X_test, y_test, verbose=0)
print(f'Test loss: {val_loss}')
print(f'Test accuracy: {val_acc}')

df_model.save('tic-tac-toe.model')

Epoch 1/50
766/766 [==============================] - 1s 958us/step - loss: 0.6459 - accuracy: 0.6475
Epoch 2/50
766/766 [==============================] - 1s 766us/step - loss: 0.6164 - accuracy: 0.6854
Epoch 3/50
766/766 [==============================] - 1s 803us/step - loss: 0.5842 - accuracy: 0.7285
Epoch 4/50
766/766 [==============================] - 1s 764us/step - loss: 0.5184 - accuracy: 0.7624
Epoch 5/50
766/766 [==============================] - 1s 778us/step - loss: 0.4840 - accuracy: 0.7872
Epoch 6/50
766/766 [==============================] - 1s 766us/step - loss: 0.4586 - accuracy: 0.8016
Epoch 7/50
766/766 [==============================] - 1s 776us/step - loss: 0.4330 - accuracy: 0.7937
Epoch 8/50
766/766 [==============================] - 1s 783us/step - loss: 0.4063 - accuracy: 0.8159
Epoch 9/50
766/766 [==============================] - 1s 806us/step - loss: 0.3858 - accuracy: 0.8264
Epoch 10/50
766/766 [==============================] - 1s 781us/step - loss: 0.358

## Step 3: Make Predictions

Now load your saved model and use it to make predictions on a few random rows in the test dataset. Check if the predictions are correct.

In [13]:
new_model = models.load_model('tic-tac-toe.model')

In [23]:
y_pred = new_model.predict(X_test[30:50])

for i, j in zip(y_test[30:50].values, y_pred):
    print(f"The actual is : {i} and the predicted is {j}")

The actual is : [0] and the predicted is [7.006526e-05]
The actual is : [1] and the predicted is [0.9997225]
The actual is : [0] and the predicted is [0.]
The actual is : [1] and the predicted is [1.]
The actual is : [1] and the predicted is [0.99999905]
The actual is : [1] and the predicted is [0.9994042]
The actual is : [1] and the predicted is [0.99066854]
The actual is : [1] and the predicted is [0.9999356]
The actual is : [1] and the predicted is [0.9997655]
The actual is : [0] and the predicted is [4.631281e-05]
The actual is : [0] and the predicted is [0.00236428]
The actual is : [1] and the predicted is [0.9976363]
The actual is : [0] and the predicted is [0.27324885]
The actual is : [0] and the predicted is [0.6535303]
The actual is : [0] and the predicted is [9.88245e-05]
The actual is : [1] and the predicted is [0.9998776]
The actual is : [0] and the predicted is [0.27589187]
The actual is : [1] and the predicted is [0.99997973]
The actual is : [0] and the predicted is [4.54

### From the above print out function we can clearly see that the predected values are equal to the actual values and that only one out of 20 values was predicted wrongly (actual = 0 / predicted = 0.6535), which proves that the accuracy of our model is 95% or in the case of the test, the accuracy was 97.9%.

## Step 4: Improve Your Model

Did your model achieve low loss (<0.1) and high accuracy (>0.95)? If not, try to improve your model.

But how? There are so many things you can play with in Tensorflow and in the next challenge you'll learn about these things. But in this challenge, let's just do a few things to see if they will help.

* Add more layers to your model. If the data are complex you need more layers. But don't use more layers than you need. If adding more layers does not improve the model performance you don't need additional layers.
* Adjust the learning rate when you compile the model. This means you will create a custom `tf.keras.optimizers.Adam` instance where you specify the learning rate you want. Then pass the instance to `model.compile` as the optimizer.
    * `tf.keras.optimizers.Adam` [reference](https://www.tensorflow.org/api_docs/python/tf/keras/optimizers/Adam).
    * Don't worry if you don't understand what the learning rate does. You'll learn about it in the next challenge.
* Adjust the number of epochs when you fit the training data to the model. Your model performance continues to improve as you train more epochs. But eventually it will reach the ceiling and the performance will stay the same.

### I tried adding layers and changing the parameters till I reached the optimal model in the cell above.
- I first tried to increase the number of units.
- I added two layers instead to two.
- I changed 'sparse_categorical_crossentropy' to 'binary_crossentropy'
- I kept changing the patch size from 20 to 2, till finally I got high accuracy with batch_size=2
- I also kept changing the number of epochs and started from 10 to 100 but noticed that with 100, I was over fitting the model and at the end I got a high accuracy with epochs=50.

**Which approach(es) did you find helpful to improve your model performance?**

In [16]:
# your answer here